proposals/format.md

```diff
@@ -0,0 +1,320 @@
+        Use(new Span<int>(args, start: 0, length: 1));
+
+        args[0] = "hello";
+        args[1] = "world";
+        Use(new Span<int>(args, start: 0, length: 2));
+
+        args[0] = "a";
+        args[1] = "longer";
+        args[2] = "set";
+        Use(new Span<int>(args, start: 0, length: 3));
+   }
+```
+
+This can siginficantly reduce the number of arrays allocated in an application. Allocations can be even further 
+reduced if the runtime provides utilities for smarter stack allocation of arrays.
+
+This optimization cannot always be applied though. Even though the callee cannot capture the `params` argument it can 
+still be captured in the caller when there is a `ref` or a `out / ref` parameter that is itself a `ref struct`
+type. 
+
+``` csharp
+static class SneakyCapture {
+    static ref int M(params Span<T> span) => ref span[0];
+
+    static void Oops() {
+        // This now holds onto the memory backing the Span<T> 
+        ref int r = ref M(42);
+    }
+}
+```
+
+These cases are statically dectable though. It potentially occurs whenever there is a `ref` return or a `ref struct`
+parameter passed by `out` or `ref`. In such a case the compiler must allocate a fresh `T[]` for every invocation. 
+
+Several other potential optimization strategies are discussed at the end of this document.
+
+The `IEnumerable<T>` variant is a merely a covenience overload. It's useful in scenarios which have frequent uses of
+`IEnumeralbe<T>` but also have lots of `params` usage. When invoked in `T` argument form the backing storage will 
+be allocated as a `T[]` just as `params T[]` is done today.
+
+### params overload resolution changes
+This proposal means the language now has four variants of `params` where before it had one. It is sensible for methods
+to define overloads of methods that differ only on the type of a `params` declarations. 
+
+Consider that `StringBuilder.AppendFormat` would certainly add a `params ReadOnlySpan<object>` overload in addition to
+the `params object[]`. This would allow it to substantially improve performance by reducing collection allocations 
+without requiring any changes to the calling code. 
+
+To facilitate this the language will introduce the following overload resolution tie breaking rule. When the candidate
+methods differ only by the `params` parameter then the canditates will be preferred in the following order:
+
+1. `ReadOnlySpan<T>`
+1. `Span<T>`
+1. `T[]`
+1. `IEnumerable<T>`
+
+This order is the most to the least effecient for the general case.
+
+### Variant
+The CoreFX is introducing a new managed type `Variant`. This type is meant to be used in APIs which expect hetrogeneous
+values but don't want the overhead brought on by using `object` as the parmeter. The `Variant` type provides universal 
+storage but avoids the boxing allocation for the most commonly used types. Using this type in APIs like `string.Format`
+can eliminate the boxing overhead in the majority of cases.
```

Comment:
is there more info on Variant and how it works?   A link, perhaps, to the CoreFx side of tings?

Reply:
I added a link to the PR that has the initial `Variant` code spit and that has a bit of conversation. I don't think there is an official design doc though. Will let @JeremyKuhne comment on that. 

https://github.com/dotnet/corefxlab/pull/2595